pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package exponentialbackoff contains logic for implementing exponential
+// backoff for GoRoutineMap and NestedPendingOperations.
+package exponentialbackoff
+
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	// initialDurationBeforeRetry is the amount of time after an error occurs
+	// that GoroutineMap will refuse to allow another operation to start with
+	// the same target (if exponentialBackOffOnError is enabled). Each
+	// successive error results in a wait 2x times the previous.
```

Comment:
This feels like a 1.4 conversation if the exponential behaviour has not
changed. I am not saying JinXu's idea don't have merit, but we do need the
fix in 1.3 and if this is how it worked in 1.2 let's decouple this
conversation.

mrubin

On Wed, Jul 20, 2016 at 10:16 AM, Jing Xu <notifications@github.com> wrote:

> In pkg/util/goroutinemap/exponentialbackoff/exponential_backoff.go
> <https://github.com/kubernetes/kubernetes/pull/28939#discussion_r71567301>
> :
>
> > +*/
> > +
> > +// Package exponentialbackoff contains logic for implementing exponential
> > +// backoff for GoRoutineMap and NestedPendingOperations.
> > +package exponentialbackoff
> > +
> > +import (
> > +	"fmt"
> > +	"time"
> > +)
> > +
> > +const (
> > +	// initialDurationBeforeRetry is the amount of time after an error occurs
> > +	// that GoroutineMap will refuse to allow another operation to start with
> > +	// the same target (if exponentialBackOffOnError is enabled). Each
> > +	// successive error results in a wait 2x times the previous.
>
> Here I am more focusing on the operations for attachable volumes. As you
> explained, if ExponentialBackoff only apply for execution error, it may not
> cause serious performance issue. But still, it may end up spending more
> time than necessary. For example, if the volume becomes attachable in 1
> min, depending on the timing, it is possible for all N pods waiting for
> more than 3 mins before starting mount operations. Also it may have higher
> chances of operation is rejected because there is an existing operation.
> What will happen if there is a hanging operation in goroutine?
>
> For volumes that are not attachable, since they can run in parallel
> without problem. Why need to run them through goroutine?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28939/files/88d495026dd3f741b1398531cbf6c484c0c03773#r71567301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASzXe9_0zvzED3kQRFsrfcDHldrKqTdnks5qXlf0gaJpZM4JMGj8>
> .
>